File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/dto/PedidoDTO.java
package br.edu.utfpr.serverpedidosjava.controller.dto;

import java.time.ZonedDateTime;
import java.util.List;

import br.edu.utfpr.serverpedidosjava.entity.Pedido;
import br.edu.utfpr.serverpedidosjava.entity.Pedido.StatusPedido;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;

public record PedidoDTO(
        Integer id,
        ZonedDateTime data,
        StatusPedido status,
        @NotNull(message = "{cliente.notnull}") ClienteDTO cliente,
        @NotNull(message = "{produtos.notempty}") @NotEmpty(message = "{produtos.notempty}") @Valid List<PedidoProdutoDTO> produtos) {

    public Pedido toEntity() {
        Pedido pedido = new Pedido();
        pedido.setId(this.id);
        pedido.setData(this.data);
        pedido.setStatus(this.status);
        pedido.setCliente(this.cliente.toEntity());
        pedido.setProdutos(this.produtos.stream().map((pedidoProduto) -> pedidoProduto.toEntity(pedido)).toList());
        return pedido;
    }

    public static PedidoDTO fromEntity(Pedido pedido) {
        return new PedidoDTO(
                pedido.getId(),
                pedido.getData(),
                pedido.getStatus(),
                ClienteDTO.fromEntity(pedido.getCliente()),
                pedido.getProdutos().stream().map((pedidoProduto) -> PedidoProdutoDTO.fromEntity(pedidoProduto))
                        .toList());
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/dto/PedidoDTO.kt
package br.edu.utfpr.serverpedidos.controller.dto

import br.edu.utfpr.serverpedidos.entity.Pedido
import br.edu.utfpr.serverpedidos.entity.StatusPedido
import jakarta.validation.Valid
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull
import java.time.ZonedDateTime

data class PedidoDTO(
    val id: Int = 0,
    val data: ZonedDateTime? = null,
    val status: StatusPedido = StatusPedido.AGUARDANDO,
    @field:NotNull(message = "{cliente.notnull}")
    val cliente: ClienteDTO? = null,
    @field:NotNull(message = "{produtos.notempty}")
    @field:NotEmpty(message = "{produtos.notempty}")
    @field:Valid
    val produtos: List<PedidoProdutoDTO>? = null
) {
    fun toEntity(): Pedido = Pedido(
        id = this.id,
        status = this.status,
        cliente = this.cliente?.toEntity()
    ).also { pedido ->
        pedido.produtos = produtos?.map { pedidoProdutoDTO ->
            pedidoProdutoDTO.toEntity(pedido)
        } ?: emptyList()
    }

    companion object {
        fun fromEntity(pedido: Pedido): PedidoDTO = PedidoDTO(
            id = pedido.id,
            data = pedido.data,
            status = pedido.status,
            cliente = pedido.cliente?.let { ClienteDTO.fromEntity(it) },
            produtos = pedido.produtos.map { PedidoProdutoDTO.fromEntity(it) }
        )
    }
}
